### PR TITLE
Yochaude/deserialization fix+refactoring

### PR DESCRIPTION
**Pull request by yochaude**
Things to consider before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [ ] Did you check UI tests on the sample app? They are not executed on CI.
* [ ] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

Fix deserialization logic for the list API. It was using the root of CosmosDB documents as opposed to the "document" object property (like everywhere else).

Also refactor and make more defensive the deserialization code. It is now shared across all API calls and unit tested in a single place.

## Related PRs or issues

List related PRs and other issues.

## Misc

Add what's missing, notes on what you tested, additional thoughts or questions.

### CODE REVIEW

File: AppCenterDataStorage/AppCenterDataStorageTests/Fixtures/testDocument.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
```
[the remark this answers is not in the document]
Comment (yochaude) [reply AppCenterDataStorage/AppCenterDataStorageTests/Fixtures/testDocument.json]: The other are pages. This one is a stand alone test document. I think `testDocument` is a pretty good name. What would you have in mind?

File: AppCenterDataStorage/AppCenterDataStorage/Internal/Util/MSDocumentUtils.h [hunk 2]
```diff
@@ -18,6 +20,31 @@ NS_ASSUME_NONNULL_BEGIN
...
 + (NSDictionary *)documentPayloadWithDocumentId:(NSString *)documentId partition:(NSString *)partition document:(NSDictionary *)document;
 
+/**
+ * Test if a reference is a dictionary that has a key of a given type.
+ *
+ * @param reference The reference to test.
+ * @param key The key to look for in the dictionary reference.
+ * @param keyType The expected key type.
+ */
++ (BOOL)isReferenceDictionaryWithKey:(id _Nullable)reference key:(NSString *)key keyType:(Class)keyType;
```
Comment (clpolet) [inline AppCenterDataStorage/AppCenterDataStorage/Internal/Util/MSDocumentUtils.h]: Would be particularly useful if we make this one a category over NSObject. We can pair on it. 
Basically importing this category you would be able to call `[myObject isReferenceDictionaryWithKey...]`.

File: AppCenterDataStorage/AppCenterDataStorageTests/Fixtures/validTestDocument.json [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  "_attachments" : "attachments/",
+  "document" : {
+    "property1" : "property number 1",
+    "property2": 123
+  },
+  "_etag" : "etag value",
+  "_ts" : 1553556495,
+  "PartitionKey" : "readonly",
+  "_rid" : "HqsVAMg0hU4EAAAAAAAAAA=="
+  }
```
Comment (dhei) [inline AppCenterDataStorage/AppCenterDataStorageTests/Fixtures/validTestDocument.json]:
```suggestion
}
```

File: AppCenterDataStorage/AppCenterDataStorageTests/MSTestDocument.h [hunk 4]
```diff
@@ -0,0 +1,21 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#import <Foundation/Foundation.h>
+
+#import "MSSerializableDocument.h"
+
+@interface MSTestDocument : NSObject <MSSerializableDocument>
+
+@property(strong, nonatomic) NSString *property1;
```
Comment (dhei) [inline AppCenterDataStorage/AppCenterDataStorageTests/MSTestDocument.h]:
```suggestion
@property(copy, nonatomic) NSString *property1;
```

File: AppCenterDataStorage/AppCenterDataStorageTests/MSTestDocument.h [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+@property(strong, nonatomic) NSNumber *property2;
```
Comment (dhei) [inline AppCenterDataStorage/AppCenterDataStorageTests/MSTestDocument.h]:
```suggestion
@property(copy, nonatomic) NSNumber *property2;
```

File: AppCenterDataStorage/AppCenterDataStorage/Internal/Util/MSDocumentUtils.h (hunk 2 above)
Comment (clpolet) [reply AppCenterDataStorage/AppCenterDataStorage/Internal/Util/MSDocumentUtils.h]: A bit overkill, at least we tried.

**Pull request merged**